src/protocol/template.rs

```diff
@@ -0,0 +1,170 @@
+            }
+
+            impl<'wire> Request<'wire> for protocol_struct!(@if ($($req_lt)?) {
+                [<$Command Request>]<'wire>
+            } else {
+                [<$Command Request>]
+            }) {
+                const TYPE: $CommandType = $CommandType::$TYPE;
+            }
+
+            impl<'wire> FromWire<'wire> for protocol_struct!(@if ($($req_lt)?) {
+                [<$Command Request>]<'wire>
+            } else {
+                [<$Command Request>]
+            }) {
+                fn from_wire<R: ReadZero<'wire> + ?Sized, A: Arena>(
+                    $r_req: &mut R,
+                    $a_req: &'wire A,
+                ) -> Result<Self, wire::Error> {
+                    $req_from
+                }
+            }
+
+            impl ToWire for protocol_struct!(@if ($($req_lt)?) {
+                [<$Command Request>]<'_>
+            } else {
+                [<$Command Request>]
+            }) {
+                fn to_wire<W: Write>(&$self_req, mut $w_req: W) -> Result<(), wire::Error> {
+                    $req_to
+                }
+            }
+
+            $(
+                make_fuzz_safe! {
+                    #[doc = "The [`" $Command "`] rspuest."]
+                    #[derive(Clone, Copy, PartialEq, Eq, Debug)]
+                    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+                    $(#[$rsp_meta])*
+                    pub $rsp_kw [<$Command Response>] $(<$rsp_lt>)? {
+                        $($rsp_fields)*
+                    }
+                }
+
+                impl<'wire> Response<'wire> for protocol_struct!(@if ($($rsp_lt)?) {
+                    [<$Command Response>]<'wire>
+                } else {
+                    [<$Command Response>]
+                }) {
+                    const TYPE: $CommandType = $CommandType::$TYPE;
+                }
+
+                impl<'wire> FromWire<'wire> for protocol_struct!(@if ($($rsp_lt)?) {
+                    [<$Command Response>]<'wire>
+                } else {
+                    [<$Command Response>]
+                }) {
+                    fn from_wire<R: ReadZero<'wire> + ?Sized, A: Arena>(
+                        $r_rsp: &mut R,
+                        $a_rsp: &'wire A,
+                    ) -> Result<Self, wire::Error> {
+                        $rsp_from
+                    }
+                }
+
+                impl ToWire for protocol_struct!(@if ($($rsp_lt)?) {
+                    [<$Command Response>]<'_>
+                } else {
+                    [<$Command Response>]
+                }) {
+                    fn to_wire<W: Write>(&$self_rsp, mut $w_rsp: W) -> Result<(), wire::Error> {
+                        $rsp_to
+                    }
+                }
+            )*
+        }
+        pub use generated::*;
+    }};
+
+    (@if () { $($t:tt)* } else { $($f:tt)* }) => {$($f)*};
+    (@if ($($nonempty:tt)+) { $($t:tt)* } else { $($f:tt)* }) => {$($t)*};
```

Comment:
Perhaps rename `@if` to something else, like `@internal`? The overlap with the `if` keyword confused me.

Reply:
`@internal if_nonempty` captures it more completely